### DISCUSSION TITLE
Digital Asset Management plugin workflow - feedback requested!

### DISCUSSION THREAD

**Discussion by ericdrosas87** [craftcms/cms#9991] [Plugin ideas] [answered]
I originally posted this in the Craft CMS Discord, but didn't get any replies so posting this here.

Hello everyone! I hope this is the right place to post this, but I am in the design stage of creating a Craft plugin that integrates with a Digital Asset Manager. The plan is to create it as vendor-agnostic as possible so we have that portability, and to open-source it/release it on the plugin store. There are actually three distinct flows represented in this diagram:

1. A daily Craft queue job that reconciles the Craft DB with the assets in the DAM
2. A content creator going into the DAM web dashboard and adding an asset, and then embedding that asset in an entry
3. Updating something about the newly added asset and having those changes asynchronously reflected in Craft world

Not pictured is the OAuth/authentication flow. I'm also hoping to not duplicate assets storage as most DAM vendors offer their own third-party storage as part of a package deal. So I'd like to create Craft DB entries for the DAM assets, but retrieve the thumbnails/URIs/metadata on the fly. 

I'm posting this here for feedback, in particular: Is my understanding of the Craft DB entries needed accurate? Are there any gotchas/showstoppers/blockers that will impede this workflow? This will be for multiple separate Craft instances (not multisite per se) integrating with a single DAM.

[Link to the Workflow](https://github.com/ericdrosas87/craft-cms-dam-integration-plugin)

**Comment by brandonkelly** [comment, marked as answer]
> I'm also hoping to not duplicate assets storage as most DAM vendors offer their own third-party storage as part of a package deal. So I'd like to create Craft DB entries for the DAM assets, but retrieve the thumbnails/URIs/metadata on the fly.

If you want to use the DAM for storage, it will also need to be used to store image transforms. Craft doesn’t currently have a way of pulling source images from one place, and storing generated transform images somewhere else.

So you will either need to create a custom volume type that implements `craft\base\VolumeInterface` entirely (including the ability for Craft to write new files to the DAM), or scrap this plan and only pull new images from the DAM, and store duplicates in a normal volume, like S3.

**Reply by ericdrosas87** [reply]
Most image transformation/editing should be occurring in the DAM itself, hence the workflow for updating of an asset also updating the DB records. Image transformations for non-DAM-bound assets will follow the usual transformation workflow.

Good to know about the `VolumeInterface` class! I'll take a look at that class in the documentation.

**Comment by jwb009** [comment]
I would love to see this come to life. We intended to integrate our DAM into Craft years ago, but it never materialized. We use OpenAsset to manage our images, so if you create a plugin that supports this you'll have a paying customer on day one!

**Reply by ericdrosas87** [reply]
I'll circle back around once the plugin is done! I'm building it for my organization's needs, but it will be open-source and available on the Craft plugin store.

**Comment by tekstrand** [comment]
@ericdrosas87 did you end up building and open sourcing this?
